pypdf/_codecs/_codecs.py

```diff
@@ -138,8 +139,123 @@ def _pack_codes_into_bytes(self, codes: List[int]) -> bytes:
 
         return bytes(output)
 
-    def decode(self, data: bytes) -> bytes:
-        """Decode data using LZW."""
-        from ..filters import LZWDecode
+    def _next_code(self, data: bytes) -> int:
+        self.bitpos: int
+        self._next_bits: int
+
+        fillbits = self.bits_per_code
+        value = 0
+
+        while fillbits > 0:
+            if self._byte_pointer >= len(data):
+                return -1
+
+            nextbits = data[self._byte_pointer]
+            bits_available = 8 - self.bitpos
+            bits_to_use = min(bits_available, fillbits)
+
+            value |= (
+                (nextbits >> (8 - self.bitpos - bits_to_use))
+                & (0xFF >> (8 - bits_to_use))
+            ) << (fillbits - bits_to_use)
+
+            fillbits -= bits_to_use
+            self.bitpos += bits_to_use
+
+            if self.bitpos == 8:
+                self.bitpos = 0
+                self._byte_pointer += 1
+
+        return value
 
-        return LZWDecode.Decoder(data).decode()
+    def next_code_decode(self) -> int:
+        self._next_data: int
+        assert self._data
+        try:
+            while self._next_bits < self._bits_to_get:
+                self._next_data = (self._next_data << 8) | (
+                    self._data[self._byte_pointer] & 0xFF
+                )
+                self._byte_pointer += 1
+                self._next_bits += 8
+
+            code = (
+                self._next_data >> (self._next_bits - self._bits_to_get)
+            ) & self._and_table[self._bits_to_get - 9]
+            self._next_bits -= self._bits_to_get
+
+            return code
+        except IndexError:
+            return 257  # End of data
+
+    def decode(self, data: bytes) -> bytes:
+        """
+        The following code was converted to Python from the following code:
+        https://github.com/empira/PDFsharp/blob/master/src/foundation/src/PDFsharp/src/PdfSharp/Pdf.Filters/LzwDecode.cs
```

Comment:
IMHO this attribution is not sufficient.

Reply:
How should I do it?

Reply:
I would use an inline comment (no docstring) for this to keep the docs clean and do it like this:

```
# The following method has been converted to Python from PDFsharp:
# https://github.com/empira/PDFsharp/blob/5fbf6ed14740bc4e16786816882d32e43af3ff5d/src/foundation/src/PDFsharp/src/PdfSharp/Pdf.Filters/LzwDecode.cs
#
# Original license:
#
# ---------------------------------------------------------------------------------
# Copyright (c) 2001-2024 empira Software GmbH, Troisdorf (Cologne Area), Germany
#
# http://docs.pdfsharp.net
#
# MIT License
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
# DEALINGS IN THE SOFTWARE.
# ---------------------------------------------------------------------------------
```

Reply:
I've added it - thank you :pray: